pyrlang/node.py

```diff
@@ -369,64 +376,136 @@ def link_nowait(self, pid1, pid2, local_only=False):
+                del to_process._links[from_pid]
+
+            return
+
+        if to_pid.is_local_to(self):  # inbound old-protocol message
+            if to_pid not in self.processes_:
+                return
+            to_process = self.processes_[to_pid]
+
+            if from_pid in to_process._links:
+                del to_process._links[from_pid]
+
+            return
+
+        # otherwise, this is an outbound request. start by performing protocol selection
+        await self._establish_connection(to_pid.node_name_)
+        conn = self.dist_nodes_.get(to_pid.node_name_, None)
+        if conn is None:
+            raise NodeException("Node %s is not connected (2)" % to_pid.node_name_)
+
+        # at this point we're done awaiting stuff, so it's safe to unpack processes
+        if from_pid not in self.processes_:
+            return
+        from_process = self.processes_[from_pid]
+        if to_pid not in from_process._links or from_process._links[to_pid] != True:
+            return
+
+        from .dist_proto.flags import DFLAG_UNLINK_ID
+        if conn.peer_flags_ & DFLAG_UNLINK_ID == 0:  # old protocol
+            del from_process._links[to_pid]
+            return await self.dist_command(to_pid.node_name_,
+                                           ('unlink', from_pid, to_pid))
+        else:  # new protocol
+            self._prev_unlink_identifier += 1
+            identifier = self._prev_unlink_identifier
+            from_process._links[to_pid] = identifier  # set to "unlinking"
+            return await self.dist_command(to_pid.node_name_,
+                                           ('unlink_id', identifier, from_pid, to_pid))
+
+    async def _unlink_id(self, identifier, from_pid, to_pid):
+        """ Services an UNLINK_ID message arriving over distribution.
+
+            NOTE: The ERTS manual says: "Upon reception of an UNLINK_ID signal,
+                  the corresponding UNLINK_ID_ACK signal must be sent before any
+                  other signals are sent to the sender of the UNLINK_ID signal."
+                  It's unclear that `await self.dist_command` satisfies this.
+        """
+
+        if to_pid not in self.processes_:
+            return
+        to_process = self.processes_[to_pid]
+        if to_process._links.get(from_pid, False) == True:
+            del to_process._links[from_pid]
+
+        # NOTE: to_pid / from_pid are deliberately swapped in the response
+        return await self.dist_command(from_pid.node_name_,
+                                       ('unlink_id_ack', identifier, to_pid, from_pid))
+
+    async def _unlink_id_ack(self, identifier, from_pid, to_pid):
+        """ Services an UNLINK_ID_ACK message arriving over distribution.
+        """
+
+        if to_pid not in self.processes_:
+            return
+        to_process = self.processes_[to_pid]
+        if to_process._links.get(from_pid, False) == identifier:
+            del to_process._links[from_pid]
```

Comment:
Sure, as you like, encapsulation won't hurt. At the same time, I would like to discourage end-users, rather than Pyrlang developers, from using these methods. What do you suggest to evoke that? Is e.g. `_set_unlink_identifier` fine, even though I'd be calling it from `node.Node`? Should I just write something scolding in the docstring? Some third option>

Reply:
`Process.add_link` has a docstring stating that `node.Node.link` should be used for proper linking.
I think something similar can be applied in this case.
Honestly for my part I would also be fine with prefixing the functions with `_` and calling them in `node.Node`
My view was mostly that of encapsulation to make it easier to alter or update the internal structure of _links.
But I'm not a Pyrlang developer per se, more of a user looking to upgrade to OTP26

Reply:
I added some accessors. It does clean up the code slightly even as-is, nevermind the possibility of better maintenance. 👍